libraries/chain/webassembly/eos-vm-oc/executor.cpp

```diff
@@ -147,20 +148,42 @@ executor::executor(const code_cache_base& cc) {
    mapping_is_executable = true;
 }
 
-void executor::execute(const code_descriptor& code, const memory& mem, apply_context& context) {
+void executor::execute(const code_descriptor& code, memory& mem, apply_context& context) {
    if(mapping_is_executable == false) {
       mprotect(code_mapping, code_mapping_size, PROT_EXEC|PROT_READ);
       mapping_is_executable = true;
    }
 
+   uint64_t max_call_depth = eosio::chain::wasm_constraints::maximum_call_depth+1;
+   uint64_t max_pages = eosio::chain::wasm_constraints::maximum_linear_memory/eosio::chain::wasm_constraints::wasm_page_size;
+   if(context.control.is_builtin_activated(builtin_protocol_feature_t::configurable_wasm_limits)) {
+      const wasm_config& config = context.control.get_global_properties().wasm_configuration;
+      max_call_depth = config.max_call_depth;
+      max_pages = config.max_pages;
+   }
+   mem.reset(max_pages);
+   EOS_ASSERT(code.starting_memory_pages <= (int)max_pages, wasm_execution_error, "Initial memory out of range");
```

Comment:
Is this line trying to detect when `config.max_pages` has been shrunk since the wasm was instantiated?

Reply:
Yes.

Reply:
Why does that check need to happen here instead of a central location?

Reply:
There isn't a convenient central location to put it.

Reply:
It seems like the check could be `wasm_interface::apply()` (by way of placing starting memory pages in the `wasm_cache_entry`). But this would incur an additional lookup every execution which would be unfortunate.